Sources/Foundation/Decimal.swift

```diff
@@ -1687,64 +1687,74 @@ public func NSDecimalString(_ dcm: UnsafePointer<Decimal>, _ locale: Any?) -> St
+    // FIXME: where magic constants 39 and 128 come from?
+    let bufferSize = 39 + 128 + 1 /* sign */ + decimalSeparatorCCharCount + 1 /* null terminator */
 
     var i = bufferSize - 1
-    while copy._exponent > 0 {
-        i -= 1
-        buffer[i] = ZERO
-        copy._exponent -= 1
-    }
-
-    if copy._exponent == 0 {
-        copy._exponent = 1
-    }
+    let buffer = Array<CChar>(unsafeUninitializedCapacity: bufferSize) { buf, initializedCount in
+        buf[bufferSize-1] = 0 // null terminator
+        while copy._exponent > 0 {
+            i -= 1
+            buf[i] = ZERO
+            copy._exponent -= 1
+        }
 
-    while copy._length != 0 {
-        var remainder: UInt16 = 0
         if copy._exponent == 0 {
+            copy._exponent = 1
+        }
+
+        while copy._length != 0 {
+            var remainder: UInt16 = 0
+            if copy._exponent == 0 {
+                i -= decimalSeparatorCCharCount
+                _ = decimalSeparatorCChar.withUnsafeBufferPointer { src in
+                    memcpy(buf.baseAddress!.advanced(by: i), src.baseAddress!, decimalSeparatorCCharCount)
+                }
+            }
+            copy._exponent += 1
+            (remainder, _) = divideByShort(&copy, 10)
             i -= 1
-            buffer[i] = DECIMALPOINT
+            buf[i] = Int8(remainder) + ZERO
         }
-        copy._exponent += 1
-        (remainder, _) = divideByShort(&copy, 10)
-        i -= 1
-        buffer[i] = Int8(remainder) + ZERO
-    }
-    if copy._exponent <= 0 {
-        while copy._exponent != 0 {
+        if copy._exponent <= 0 {
+            while copy._exponent != 0 {
+                i -= 1
+                buf[i] = ZERO
+                copy._exponent += 1
+            }
+            i -= decimalSeparatorCCharCount
+            _ = decimalSeparatorCChar.withUnsafeBufferPointer { src in
+                memcpy(buf.baseAddress!.advanced(by: i), src.baseAddress!, decimalSeparatorCCharCount)
+            }
             i -= 1
-            buffer[i] = ZERO
-            copy._exponent += 1
+            buf[i] = ZERO
         }
-        i -= 1
-        buffer[i] = DECIMALPOINT
-        i -= 1
-        buffer[i] = ZERO
-    }
-    if copy._isNegative != 0 {
-        i -= 1
-        buffer[i] = MINUS
-    }
-    let result = String(cString: Array(buffer.suffix(from: i)))
-
-    var decimalSeparator: String?
-    if let locale = locale as? Locale {
-        decimalSeparator = locale.decimalSeparator
-    } else if let dictionary = locale as? [NSLocale.Key: String] {
-        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator]
-    } else if let dictionary = locale as? [String: String] {
-        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator.rawValue]
+        if copy._isNegative != 0 {
+            i -= 1
+            buf[i] = MINUS
+        }
+        memset(buf.baseAddress, 0, i) // FIXME: this could be avoided as well
```

Comment:
Unfortunately the shorter the final result the more bytes here are getting initialised but not really needed. Having a heuristic that tries to overestimate the final string length would help reduce the work done here.